query.go

```diff
@@ -71,67 +85,103 @@ func (q *dhtQuery) Run(ctx context.Context, peers []peer.ID) (*dhtQueryResult, e
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	runner := newQueryRunner(q)
+	runner := newQueryRunner(q, numPaths)
 	return runner.Run(ctx, peers)
 }
 
-type dhtQueryRunner struct {
-	query          *dhtQuery        // query to run
-	peersSeen      *pset.PeerSet    // all peers queried. prevent querying same peer 2x
-	peersQueried   *pset.PeerSet    // peers successfully connected to and queried
+type dhtQueryPath struct {
+	runner         *dhtQueryRunner
+	qfunc          QueryFunc
 	peersToQuery   *queue.ChanQueue // peers remaining to be queried
 	peersRemaining todoctr.Counter  // peersToQuery + currently processing
 
 	result *dhtQueryResult // query result
-	errs   u.MultiErr      // result errors. maybe should be a map[peer.ID]error
 
 	rateLimit chan struct{} // processing semaphore
-	log       logging.EventLogger
+	proc      process.Process
+}
+
+type dhtQueryRunner struct {
+	query          *dhtQuery // query to run
+	paths          []*dhtQueryPath
+	pathsRemaining todoctr.Counter
+	peersSeen      *pset.PeerSet // all peers queried. prevent querying same peer 2x
+	peersQueried   *pset.PeerSet // peers successfully connected to and queried
+
+	errs u.MultiErr // result errors. maybe should be a map[peer.ID]error
+	log  logging.EventLogger
 
 	runCtx context.Context
+	result *dhtQueryFullResults
 
 	proc process.Process
 	sync.RWMutex
```

Comment:
While moving this struct around please use the [mutex hat](https://dmitri.shuralyov.com/idiomatic-go#mutex-hat) pattern to make clear what the mutex is protecting.

Reply:
Do you mean that anything protected by the RWMutex should be at the end? That isn't how the code was before so I'm slightly confused.

Reply:
So anything protected by the mutex should be directly underneath the mutex declaration and there should be a newline before any variables that aren't protected by the mutex, i.e.:
```go
 struct {
 	...

 	rateMu     sync.Mutex
 	rateLimits [categories]Rate          // protected by rateMu
 	mostRecent rateLimitCategory    // protected by rateMu
+
+	common service      // NOT protected by rateMu
 }
```

Reply:
@jhiesey yep and remove any newlines between the fields so its clear it is protecting everything :+1: